### DISCUSSION TITLE
How to: Local storage on Browser? Bookmarks?

### DISCUSSION THREAD

**Discussion by andrewkdci** [AvaloniaUI/Avalonia#15487] [Q&A]
I need to provide some local storage on the Browser, for settings files that will be specific to each client. It just needs to be a simple text file(s).

I am looking into the [StorageProvider](https://docs.avaloniaui.net/docs/concepts/services/storage-provider/) via [TopLevel](https://docs.avaloniaui.net/docs/concepts/toplevel), particularly [Bookmarks](https://docs.avaloniaui.net/docs/concepts/services/storage-provider/bookmarks), but I'm unable to get it to work. I can't seem to figure out how to create a bookmark let alone read one.

I see that StorageProvider has OpenFolderBookmarkAsync and OpenFileBookmarkAsync methods, but no creation methods. Still, I tried those but the resulting IStorageBookFolder returned is null. Furthermore, I don't know what string to put in for the bookmark. Is it any unique string, or does it have to be a path?

Bonus if there's a solution that works on both Browser and Desktop (Windows and Linux). However I can accept having to write platform-specific code as long as I can get local storage to work on all platforms.

**Comment by thevortexcloud** [comment]
It looks like it will return null if no book mark is found:

https://github.com/AvaloniaUI/Avalonia/blob/acdabef81ac35563da11ce56ae7a65ee2fa4fe97/src/Browser/Avalonia.Browser/Storage/BrowserStorageProvider.cs#L105-L117

`OpenBookmark` itself just calls some Javascript API.

https://github.com/AvaloniaUI/Avalonia/blob/acdabef81ac35563da11ce56ae7a65ee2fa4fe97/src/Browser/Avalonia.Browser/Interop/StorageHelper.cs#L25-L26

Looking at the Javascript for `OpenBookmark` it looks like it uses some database.

https://github.com/AvaloniaUI/Avalonia/blob/acdabef81ac35563da11ce56ae7a65ee2fa4fe97/src/Browser/Avalonia.Browser/webapp/modules/storage/storageProvider.ts#L59-L67

To save a bookmark, it looks like you just need to call SaveBookmark

https://github.com/AvaloniaUI/Avalonia/blob/acdabef81ac35563da11ce56ae7a65ee2fa4fe97/src/Browser/Avalonia.Browser/Storage/BrowserStorageProvider.cs#L191-L199


The API seems to be implemented on desktops too.

**Reply by maxkatz6** [reply]
That's correct.
Typical workflow is:
1. Open file/folder via file picker.
2. Bookmark this file/folder by getting a string identifier. On Browser specifically SaveBookmarkAsync will save file reference to the IndexedDB, and return key of that. 
3. Save this identifier as you prefer. On browser it would likely be localStorage HTML API.
4. When (web) app is opened next time, you can skip file picker, and just load bookmark by its identifier.
